### PR TITLE
Labwc window border color support

### PR DESCRIPTION
**Pull request by dmitry-erin**
This PR adds patch for labwc, which allows to change the frame's color for chosen app using window rules.

## Description of changes

The labwc patch allows to change the frame's color for chosen app using window rules. This patch was added to default.nix for labwc overlay. The task number is SP-2806. The better solution will be discussed here https://github.com/labwc/labwc/discussions/1092.

## Checklist for things done

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [x] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [x] PR linked to architecture documentation and requirement(s) (ticket id)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `nix flake check --accept-flake-config` and it passes
- [x] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

## Testing

Uncomment this line  `# profiles.graphics.compositor = "labwc";` in quivm.nix to enable labwc. Build debug target for Lenovo X1, run it, launch terminal. The terminal will have the aqua colored frame.

### CODE REVIEW

**Comment by vilvo** [comment]
Please fill the PR template by checking the relevant checkboxes. Also remove the "Draft" status when the PR is ready for review.

Some early commits.
Please squash the commits into one. Commit 5975f5d is not relevant to store in `ghaf/main` version history is it's only your dev time formatting change (important) but should be squashed into commit with other work. 
Then you can mark this checkbox item as done: "Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR"

Bigger issues is in the PR patch content:
- your `.patch` must not change `flake.lock` - 
please remove these https://github.com/tiiuae/ghaf/pull/416/files#diff-cb3638f781067be81fe52f5164a8bff0bf1da1a554c72c7a4d3aace3b1e85b59R56-R122
- your `.patch` must not patch `flake.nix` this way https://github.com/tiiuae/ghaf/pull/416/files#diff-cb3638f781067be81fe52f5164a8bff0bf1da1a554c72c7a4d3aace3b1e85b59R123-R156 - integrators will not allow this. 
  - Please refactor your labwc packaging as follows:
    - create `labwc`-directory under https://github.com/tiiuae/ghaf/tree/main/packages
    - move these changes https://github.com/tiiuae/ghaf/pull/416/files#diff-cb3638f781067be81fe52f5164a8bff0bf1da1a554c72c7a4d3aace3b1e85b59R9-R55 into `default.nix` under directory from previous step
    - do not do this https://github.com/tiiuae/ghaf/pull/416/files#diff-cb3638f781067be81fe52f5164a8bff0bf1da1a554c72c7a4d3aace3b1e85b59R144-R150 in the top-level flake but in this file instead https://github.com/tiiuae/ghaf/blob/main/packages/default.nix
    
These [`labwc`-patching changes seem problematic](https://github.com/tiiuae/ghaf/pull/416/files#diff-cb3638f781067be81fe52f5164a8bff0bf1da1a554c72c7a4d3aace3b1e85b59R157-R669) but I did not test them yet. To me it seems that it assumes that `labwc` repository is cloned manually to somewhere inside ghaf-repository so that these apply. All of that cloning should go to labwc nix-packaging and then the patching would take place in that packaging but that's not in place in this PR. Did I miss something?

**Review by vilvo — CHANGES_REQUESTED** [review]
Please see details of requested changes in https://github.com/tiiuae/ghaf/pull/416#issuecomment-1862385105

**Comment by vilvo** [comment]
Would you please mind changing this very long PR topic to short summary? 
screencapture
I propose: "Labwc window border color support"

File: docs/src/ref_impl/labwc_config.md
Comment (vilvo) [inline docs/src/ref_impl/labwc_config.md:5]: Please add here a subsection "#Overview" or "#Rationale" that summarizes why this concept has been developed.

A few keys to help you:
- border color concept illustrates the app trustworthiness
- please attribute earlier work of the concept to https://www.qubes-os.org/doc/getting-started/#color--security (possible other before QubesOS)
- list the limitations of current implementation - e.g. if the trustworthiness color is forgeable and other things you see that need further development

File: modules/virtualization/microvm/guivm.nix
Comment (vilvo) [inline modules/virtualization/microvm/guivm.nix:24]: We can't yet this to default to "labwc" so let's keep it just configurable to "labwc" until further development. We use the same - not default to new features - with other security features under development as well.

File: docs/src/ref_impl/labwc_config.md
Comment (vilvo) [reply docs/src/ref_impl/labwc_config.md:5]: This part of the documentation would also benefit of screencapture image.

File: overlays/custom-packages/labwc/default.nix
Comment (vilvo) [inline overlays/custom-packages/labwc/default.nix:8]: We can do this labwc patching in `packages/labwc/default.nix` instead.

File: overlays/custom-packages/labwc/default.nix
Comment (vilvo) [inline overlays/custom-packages/labwc/default.nix:31]: The same applies here as well - modifying labwc per app trustworthiness color config. If you move this to `packages\labwc\default.nix`, we can drop this overlay and use that labwc package directly instead.

**Review by vilvo — CHANGES_REQUESTED** [review]
Lot of great structural improvements since the last review request.
Please address the documentation improvement and a few more structural improvements requests from review. 

I did not yet throughly review the labwc `.patch` but there was some `hex_to_dec` and `parse_hexstr` code refactoring from source file to another(?) that raises questions. These kind of changes may be hard to maintain outside upstream. I think it could make sense to work those parts of this concept example to the level where the review and comments on coloring the borders could be requested directly from the upstream labwc-project. When properly implemented, that part might be accepted to the labwc-upstream project where it would be best maintained. This is probably not a ghaf concept demo blocker, though.

Let's leave above labwc-patch discussion/decision until the next/final review round.

**Comment by dmitry-erin** [comment]
Rolled back to overlay only, think about making a package as proposed. Previous changes wasn't tested properly.

File: modules/virtualization/microvm/guivm.nix
Comment (unbel13ver) [reply modules/virtualization/microvm/guivm.nix:24]: I agree - we need to make sure everything (and everyone) is ready to switch to `labwc` first.

File: overlays/custom-packages/labwc/default.nix
Comment (unbel13ver) [reply overlays/custom-packages/labwc/default.nix:8]: @vilvo I am not sure I understand what is `packages/labwc/default.nix` that you mentioned. In my opinion, this overlay seems to be the best place for keeping the patch.

File: docs/src/ref_impl/labwc_config.md
Comment (unbel13ver) [inline docs/src/ref_impl/labwc_config.md:2]: I believe you need to put 2024 here from now on =)

**Review by unbel13ver — CHANGES_REQUESTED** [review]

File: docs/src/ref_impl/labwc_config.md
Comment (vilvo) [inline docs/src/ref_impl/labwc_config.md:1]: The basic technical content is there in this file but please work with @jenninikko to:
- improve the English
- make the structure more consistent with other docs

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:174]: The code indentation here on lines 174-176 is non-consistent which kind of adds to hackiness of the concept patch. Please indent.

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:197]: Same here, please see previous comment.

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:226]: Please remove commented out code.

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:306]: See previous comment. If you want to leave some marks on the 2nd approach, please discuss in the markdown-document what was tried and why it did not work. Or why you ended up with the 1st approach: "the whole theme changing"

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:20]: The common practice to extend (or hack) with additional parameters is to add them to the end of the parameter list - like to add `float *bg_color` in the end, not in the middle. The same applies on extending enums, structs.
If you have a specific reason to add it randomly in the middle, please leave a code comment.

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:253]: https://github.com/tiiuae/ghaf/pull/416/files#r1442600464

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:269]: https://github.com/tiiuae/ghaf/pull/416/files#r1442600464

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:345]: Indentation.

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:362]: Was it absolutely necessary to move `static [hex_to_dec|parse_hexstr]` from `.c` to `.h`?
Could you just declare them in the header and leave the implementation in `.c` files?

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:462]: Pleas remove commented out code.

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (vilvo) [inline overlays/custom-packages/labwc/labwc-colored-borders.patch:472]: ??

**Review by vilvo — CHANGES_REQUESTED** [review]
Please address the `.patch` review comments. Nothing blocking but to improve the border color hack patch for future readers.
In addition - I had a chat with @unbel13ver and heard you had discussed the border coloring feature with the labwc upstream maintainer. I understand that's a longer term approach but as a minimum requirement please document thoroughly in this git commit message the rationale for this concept patch and what has been discussed with the labwc (preferably with links to upstream project forums where) on having this as an upstream feature. This is critical documentation because I see the current patch a maintenance challenge with labwc updates. I would be surprised if the patch would not cause build breaks with updates which means the you or other people will have to look at the patch, understand the context and maintain it.

**Comment by vilvo** [comment]
> a minimum requirement please document thoroughly in this git commit message 

Make sure the information goes into *git commit message* so that the information stays with the code changes for future readers.

File: overlays/custom-packages/labwc/labwc-colored-borders.patch
Comment (dmitry-erin) [reply overlays/custom-packages/labwc/labwc-colored-borders.patch:20]: The function 
```
struct ssd_part *
add_scene_button(struct wl_list *part_list, enum ssd_part_type type,
		struct wlr_scene_tree *parent, float *bg_color,
		struct wlr_buffer *icon_buffer, struct wlr_buffer *hover_buffer,
		int x, struct view *view)
```
has _bg_color_  parameter and is used in the 

```
struct ssd_part *
add_scene_button_corner(struct wl_list *part_list, enum ssd_part_type type,
		enum ssd_part_type corner_type, struct wlr_scene_tree *parent,
		struct wlr_buffer *corner_buffer, struct wlr_buffer *icon_buffer,
		struct wlr_buffer *hover_buffer, int x, struct view *view)
```
I just tried to unify the parameters ordering. It helps avoiding mistakes.

Comment (vilvo) [reply overlays/custom-packages/labwc/labwc-colored-borders.patch:20]: Thanks for clarifying your thinking here. If I'm not mistaken, the parameter ordering comes from the upstream project. How could we check if this contribution would be accepted to the upstream project for easier maintenance?

Comment (dmitry-erin) [reply overlays/custom-packages/labwc/labwc-colored-borders.patch:20]: I asked the maintainers and waiting for an answer. This solution doesn't change the overall logic (but must do it) and just substitutes the colors in a few code snippets. To introduce the colored window borders as a feature the theme handling must be revised (in my opinion).

**Review by jenninikko — APPROVED** [review]

**Comment by dmitry-erin** [comment]
To https://github.com/tiiuae/ghaf/pull/416#pullrequestreview-1811144992
Not sure it would be useful, there is not a lot of info in this discussion. I mentioned in my previous comment (https://github.com/tiiuae/ghaf/pull/416#discussion_r1444251569) that a "well" approach must include revised theme handling logic. The main "rationale" of using patch is that I have no "well" solution yet. Additionally such a big changes will definitely lead to the regression...

**Review by jenninikko — APPROVED** [review]

**Comment by leivos-unikie** [comment]
Changed default compositor to "labwc" in modules/profiles/graphics.nix, built the image and tested.
- Couldn't launch demo apps (no icons). 
- Terminal launches by clicking left mouse button and selecting 'Terminal'. Indeed, terminal window has borders with aqua color.
- wifi-connector works
- ci-test-automation run passes (except launching chromium)

**Comment by vilvo** [comment]
> To [#416 (review)](https://github.com/tiiuae/ghaf/pull/416#pullrequestreview-1811144992) Not sure it would be useful, there is not a lot of info in this discussion. I mentioned in my previous comment ([#416 (comment)](https://github.com/tiiuae/ghaf/pull/416#discussion_r1444251569)) that a "well" approach must include revised theme handling logic. The main "rationale" of using patch is that I have no "well" solution yet. Additionally such a big changes will definitely lead to the regression...

Thanks for adding https://github.com/labwc/labwc/discussions/1092 to the git commit.

**Review by vilvo — APPROVED** [review]
It's a hack but I got the aqua color terminal in testing. 
I'm looking forward to your constructive work and collaboration with upstream to replace this.

**Comment by vilvo** [comment]
Please update "Testing" section of the PR to mention that user must enable "labwc" in guivm for testing.
This is also worth mentioning in the git commit (with `--amend`) 
```
diff --git a/modules/virtualization/microvm/guivm.nix b/modules/virtualization/microvm/guivm.nix
index 028b6b9..46acf3d 100644
--- a/modules/virtualization/microvm/guivm.nix
+++ b/modules/virtualization/microvm/guivm.nix
@@ -21,7 +21,7 @@
           users.accounts.enable = lib.mkDefault configHost.ghaf.users.accounts.enable;
           profiles.graphics.enable = true;
           # Uncomment this line to take LabWC in use
-          # profiles.graphics.compositor = "labwc";
+          profiles.graphics.compositor = "labwc";
```
I knew I had to enable it but as I haven't tested labwc lately so I had to search around ghaf repo for it.

**Review by vilvo — APPROVED** [review]

**Review by brianmcgillion — APPROVED** [review]

**Pull request merged**
